File: app/src/main/java/com/lovejjfg/arsenal/ui/ArsenalSearchActivity.java
package com.lovejjfg.arsenal.ui;

import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.transition.TransitionManager;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.lovejjfg.arsenal.R;
import com.lovejjfg.arsenal.base.SupportActivity;
import com.lovejjfg.arsenal.utils.JumpUtils;
import com.lovejjfg.arsenal.utils.TagUtils;
import com.lovejjfg.arsenal.utils.rxbus.RxBus;
import com.lovejjfg.arsenal.utils.rxbus.SearchEvent;
import com.miguelcatalan.materialsearchview.MaterialSearchView;

import butterknife.Bind;
import butterknife.ButterKnife;

public class ArsenalSearchActivity extends SupportActivity implements View.OnClickListener {
    private static final String TAG = ArsenalSearchActivity.class.getSimpleName();
    @Bind(R.id.toolbar)
    Toolbar mToolbar;
    @Bind(R.id.search_view)
    MaterialSearchView searchView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ButterKnife.bind(this);
        setSupportActionBar(mToolbar);
        String currentTag = getIntent().getStringExtra(ArsenalListInfoFragment.KEY);
        String currentTagName = getIntent().getStringExtra(ArsenalListInfoFragment.TAG_NAME);
        getSupportActionBar().setTitle(TextUtils.isEmpty(currentTagName) ? currentTag : currentTagName);
        mToolbar.setNavigationOnClickListener(this);

        if (savedInstanceState == null) {
            ArsenalListInfoFragment listInfoFragment = new ArsenalListInfoFragment();
            Bundle bundle = new Bundle();
            bundle.putInt(ArsenalListInfoFragment.TYPE_NAME, getIntent().getIntExtra(ArsenalListInfoFragment.TYPE_NAME, ArsenalListInfoFragment.TYPE_SEARCH_TAG));
            bundle.putString(ArsenalListInfoFragment.KEY, currentTag);
            listInfoFragment.setArguments(bundle);
            loadRoot(R.id.activity_tag_search, listInfoFragment);
        }
//        setSupportActionBar(mToolbar);

        searchView.setVoiceSearch(false);
        searchView.setOnSearchViewListener(new MaterialSearchView.SearchViewListener() {
            @Override
            public void onSearchViewShown() {
                String[] strings = TagUtils.getTagArray();
                if (strings != null) {
                    searchView.setSuggestions(strings, new MaterialSearchView.SuggestionsListCallBack() {
                        @Override
                        public void onItemClick(String title) {
                            String s = TagUtils.getTagValue(title);
                            JumpUtils.jumpToSearchList(searchView.getContext(), title, "/tag/" + s, ArsenalListInfoFragment.TYPE_SEARCH_TAG);
                            searchView.closeSearch();
                        }
                    });
                }
            }

            @Override
            public void onSearchViewClosed() {
                //Do some magic
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                    TransitionManager.beginDelayedTransition((ViewGroup) mToolbar.getParent());
                }
                mToolbar.setVisibility(View.VISIBLE);
            }

            @Override
            public void onSearchViewAnimationStart() {
                mToolbar.setVisibility(View.GONE);
            }
        });

        searchView.setEllipsize(true);

        searchView.setOnQueryTextListener(new MaterialSearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                Log.e(TAG, "onQueryTextSubmit: " + query);
                searchView.closeSearch();
                JumpUtils.jumpToSearchList(ArsenalSearchActivity.this, query, ArsenalListInfoFragment.TYPE_SEARCH);
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                //Do some magic
                return false;
            }
        });


    }

    @Override
    public int initLayoutRes() {
        return R.layout.activity_tag_search;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_search, menu);
        final MenuItem myActionMenuItem = menu.findItem(R.id.action_search);
        searchView.setMenuItem(myActionMenuItem);
        return true;
    }

    @Override
    public void onBackPressed() {
        if (searchView.isSearchOpen()) {
            searchView.closeSearch();
            return;
        }
        super.onBackPressed();
    }

    @Override
    public void onClick(View v) {
        onBackPressed();
    }

    @Override
    protected void onNewIntent(Intent intent) {
        String currentTag = intent.getStringExtra(ArsenalListInfoFragment.KEY);
        String currentTagName = intent.getStringExtra(ArsenalListInfoFragment.TAG_NAME);
        int type = intent.getIntExtra(ArsenalListInfoFragment.TYPE_NAME, ArsenalListInfoFragment.TYPE_SEARCH_TAG);
        getSupportActionBar().setTitle(TextUtils.isEmpty(currentTagName) ? currentTag : currentTagName);
        RxBus.getInstance().post(new SearchEvent(currentTag, currentTagName, type));
        super.onNewIntent(intent);
    }
}

File: app/src/main/java/com/lovejjfg/arsenal/ui/ArsenalSearchActivity.kt
/*
 *  Copyright (c) 2017.  Joe
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

package com.lovejjfg.arsenal.ui

import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.animation.BounceInterpolator
import com.lovejjfg.arsenal.R
import com.lovejjfg.arsenal.api.mode.SearchInfo
import com.lovejjfg.arsenal.base.SupportActivity
import com.lovejjfg.arsenal.utils.FirebaseUtils
import com.lovejjfg.arsenal.utils.JumpUtils
import com.lovejjfg.arsenal.utils.TagUtils
import com.lovejjfg.arsenal.utils.rxbus.RxBus
import com.lovejjfg.arsenal.utils.rxbus.SearchEvent
import com.lovejjfg.arsenal.utils.transationHint
import com.miguelcatalan.materialsearchview.MaterialSearchView
import kotlinx.android.synthetic.main.activity_tag_search.activity_tag_search
import kotlinx.android.synthetic.main.activity_tag_search.search_view

class ArsenalSearchActivity : SupportActivity(), View.OnClickListener {
    lateinit var searchView: MaterialSearchView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        searchView = search_view
        setSupportActionBar(toolbar)
        val currentTag = intent.getStringExtra(ArsenalListInfoFragment.KEY)
        val currentTagName = intent.getStringExtra(ArsenalListInfoFragment.TAG_NAME)
        supportActionBar?.title = if (TextUtils.isEmpty(currentTagName)) currentTag else currentTagName
        toolbar?.setNavigationOnClickListener(this)

        if (savedInstanceState == null) {
            val listInfoFragment = ArsenalListInfoFragment()
            val bundle = Bundle()
            bundle.putInt(
                ArsenalListInfoFragment.TYPE_NAME, intent.getIntExtra(
                    ArsenalListInfoFragment.TYPE_NAME,
                    ArsenalListInfoFragment.TYPE_SEARCH_TAG
                )
            )
            bundle.putString(ArsenalListInfoFragment.KEY, currentTag)
            listInfoFragment.arguments = bundle
            loadRoot(R.id.activity_tag_search, listInfoFragment)
        }
        //        setSupportActionBar(mToolbar);

        searchView.setVoiceSearch(false)
        searchView.setOnSearchViewListener(object : MaterialSearchView.SearchViewListener {
            override fun onSearchViewShown() {
                val strings = TagUtils.tagArray
                if (strings != null) {
                    searchView.setSuggestions(strings) { title ->
                        if (searchView.isSearchOpen) {
                            searchView.closeSearch()
                            val s = TagUtils.getTagValue(title)
                            if (TextUtils.isEmpty(s)) {
                                JumpUtils.jumpToSearchList(searchView.context, title)
                            } else {
                                JumpUtils.jumpToSearchList(searchView.context, title, "/tag/" + s!!)
                            }
                        }
                    }
                }
            }

            override fun onSearchViewClosed() {
                //Do some magic
                //                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                //                    TransitionManager.beginDelayedTransition((ViewGroup) mToolbar.getParent(), new Fade(Fade.IN));
                //                }
                //                mToolbar.setVisibility(View.VISIBLE);
            }

            override fun onSearchViewAnimationStart() {
                //                mToolbar.setVisibility(View.INVISIBLE);
            }
        })

        searchView.setEllipsize(true)

        searchView.setOnQueryTextListener(object : MaterialSearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                searchView.closeSearch()
                val searchInfo = SearchInfo(query, ArsenalListInfoFragment.TYPE_SEARCH, null)
                TagUtils.save(searchInfo)
                JumpUtils.jumpToSearchList(this@ArsenalSearchActivity, query)
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                //Do some magic
                return false
            }
        })
    }

    override fun initLayoutRes(): Int {
        return R.layout.activity_tag_search
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_search, menu)
        //        final MenuItem myActionMenuItem = menu.findItem(R.id.action_search);
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_search -> {
                if (searchView.menuItem == null) {
                    val searchMenuView = toolbar?.findViewById<View>(R.id.action_search)
                    searchView.menuItem = searchMenuView
                }
                if (!searchView.isSearchOpen) {
                    searchView.showSearch()
                }
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        if (searchView.isSearchOpen) {
            searchView.closeSearch()
            return
        }
        super.onBackPressed()
    }

    override fun onClick(v: View) {
        onBackPressed()
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        val currentTag = intent.getStringExtra(ArsenalListInfoFragment.KEY)
        val currentTagName = intent.getStringExtra(ArsenalListInfoFragment.TAG_NAME)
        val type = intent.getIntExtra(
            ArsenalListInfoFragment.TYPE_NAME,
            ArsenalListInfoFragment.TYPE_SEARCH_TAG
        )
        val title = supportActionBar?.title
        println("title:$title;;currentTag: $currentTag || currentTagName: $currentTagName")
        if (title == currentTag || title == currentTagName) {
            activity_tag_search.transationHint()
        } else {
            val t = if (TextUtils.isEmpty(currentTagName)) currentTag else currentTagName
            supportActionBar?.title = t
            RxBus.getInstance().post(SearchEvent(currentTag, currentTagName, type))
            FirebaseUtils.logSearchEvent(this@ArsenalSearchActivity,t)
        }
    }
}
